### PR TITLE
Add RRT motion planner which can integrate into different robots

### PR DESCRIPTION
**Pull request by cpaxton**
## Motivation and Context

- Integrate into 2d/3d maps
- Use this to make sure the robot isn't hitting stuff

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

# Checklist:

- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] I ran `tests/hw_manual_test.py` on hardware and observed that robot behavior looks normal.

### CODE REVIEW

File: src/home_robot/home_robot/mapping/voxel.py [hunk 1]
```diff
@@ -194,6 +206,14 @@ def add(
             assert depth.ndim == 2
         if camera_K is not None:
             assert camera_K.ndim == 2
+        if instance_image is None:
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:209]: Nice error checking

File: src/home_robot/home_robot/mapping/voxel.py [hunk 2]
```diff
@@ -194,6 +206,14 @@ def add(
...
+            assert (
+                obs is not None
+            ), "must provide instance image or raw observations with instances"
+            assert (
+                obs.instance is not None
+            ), "must provide instance image in observation if not available otherwise"
+            instance_image = torch.from_numpy(obs.instance)
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:216]: Should we have a wrapper method add_obs(obs) that unpacks obs and calls add?

File: src/home_robot/home_robot/mapping/voxel.py [hunk 3]
```diff
@@ -492,18 +552,65 @@ def _show_pytorch3d(self, **plot_scene_kwargs):
...
+            np.bitwise_and(
+                np.bitwise_not(obstacles.cpu().numpy()), explored.cpu().numpy()
+            )
+            == 1
+        )
+        if len(valid_indices) > 0:
+            random_index = np.random.choice(len(valid_indices))
+            return self.grid_coords_to_xy(valid_indices[random_index])
+        else:
+            return None
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:568]: This function can probably be written in torch and then cast to numpy at the last second

File: src/home_robot/home_robot/mapping/voxel.py [hunk 4]
```diff
@@ -424,6 +463,27 @@ def get_2d_map(
...
+        """convert grid coordinate point to metric world xy point"""
+        assert grid_coords.shape[-1] == 2, "grid coords must be an Nx2 or 2d array"
+        # TODO: tensor version of this
+        return (grid_coords - self.grid_origin[:2].cpu().numpy()) * self.grid_resolution
+
+    def grid_coords_to_xyt(self, grid_coords: np.ndarray) -> np.ndarray:
+        """convert grid coordinate point to metric world xyt point"""
+        res = np.zeros(3)
+        res[:2] = self.grid_coords_to_xy(grid_coords)
+        return res
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:485]: These functions can all be written in torch

File: src/home_robot/home_robot/utils/point_cloud_torch.py [hunk 5]
```diff
@@ -247,4 +247,10 @@ def get_bounds(points: Tensor):
     Returns:
         mins_and_maxes: [3, 2]
     """
-    return torch.stack([points.min(dim=0)[0], points.max(dim=0)[0]], dim=-1)
+    mins = points.min(dim=0)[0]
+    maxs = points.max(dim=0)[0]
+    if not torch.all(maxs > mins):
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/point_cloud_torch.py:252]: Maybe add a `warnings.warn(RuntimeError("Zero-volume bounds detected in get_bounds"))`

File: src/home_robot/home_robot/utils/simple_env.py [hunk 6]
```diff
@@ -5,25 +5,80 @@
...
+        return True
 
-    @abstractmethod
     def reset(self, seed: int = 0):
         self.seed = seed
         np.random.seed(seed)
-        # TODO: regenerate start, goal and obstacles
+        self.obstacle_pos = np.random.random(2) * (self.size - self.obstacle_size)
+
+    def show(self, states):
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/simple_env.py:55]: Can we start to implement backends, so
`def show(self, states, backend='mpl', show=True)`
Then the function by default returns a figure and only calls plt.show() if show == True

**Review by alexsax — CHANGES_REQUESTED** [review]
Generally LGTM but for functions that are essentially the same in torch (e.g. the new methods in SparseVoxelMap/NavigationSpace), can we move to doing them in torch? 
As a side-benefit np.bitwise_and or logical_or end up cleaner like (a & b) or (a | b)

File: src/home_robot/home_robot/utils/simple_env.py (hunk 6 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/simple_env.py:55]: done

File: src/home_robot/home_robot/utils/point_cloud_torch.py (hunk 5 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/point_cloud_torch.py:252]: done

File: src/home_robot/home_robot/mapping/voxel.py (hunk 2 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:241]: done

File: src/home_robot/home_robot/mapping/voxel.py (hunk 3 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py]: fixed

File: src/home_robot/home_robot/mapping/voxel.py (hunk 4 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py]: fixed

**Comment by cpaxton** [comment]
@hello-binit adding RRT motion planner for safe trajectory execution when exploring

File: src/home_robot/home_robot/mapping/voxel.py (hunk 1 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel.py:236]: Thanks!

File: src/home_robot/home_robot/mapping/voxel.py [hunk 7]
```diff
@@ -139,6 +155,28 @@ def get_instances(self) -> List[InstanceView]:
         """Return a list of all viewable instances"""
         return tuple(self.instances.instance_views[0].values())
 
+    def add_obs(self, obs: Observations, *args, **kwargs):
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:158]: Beautiful

File: src/home_robot/home_robot/mapping/voxel.py [hunk 8]
```diff
@@ -312,6 +365,17 @@ def write_to_pickle(self, filename: str):
         with open(filename, "wb") as f:
             pickle.dump(data, f)
 
+    def fix_data_type(self, tensor) -> torch.Tensor:
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel.py:368]: I like this method. Small nit: I'd probably just call this `ensure_float_tensor`

File: src/home_robot/home_robot/mapping/voxel_map.py [hunk 9]
```diff
@@ -2,25 +2,329 @@
...
+            valid_indices = torch.nonzero(expanded_frontier, as_tuple=False)
+            if valid_indices.size(0) == 0:
+                continue
+
+            # Rejection sampling:
+            # - Find a point that we could potentially move to
+            # - Compute a position and orientation
+            # - Check to see if we can actually move there
+            # - If so, return it
+            for i in range(max_tries_per_size):
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel_map.py:280]: I think this is reasonable and it's good to have a first version that works. 
If RRT becomes slow then this rejection sampling should be easy to parallelize/run on GPU

File: src/home_robot/home_robot/utils/morphology.py [hunk 10]
```diff
@@ -49,3 +50,103 @@ def binary_closing(binary_image, kernel):
...
     return binary_opening(binary_closing(binary_image, kernel), kernel)
+
+
+def get_edges(mask: torch.Tensor, threshold: float = 0.5) -> torch.Tensor:
+    """Extract edges from a torch tensor."""
+
+    mask = mask.float()
+
+    # Define the Sobel filter kernels
+    sobel_x = torch.tensor([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]], dtype=torch.float32)
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/morphology.py:61]: If `mask` is on gpu then this will cause an error unless the kernels are on the same device: i.e.
`sobel_x = torch.tensor([[-1, 0, 1], [-2, 0, 2], [-1, 0, 1]], dtype=torch.float32, device=mask.device)`

File: src/home_robot/home_robot/utils/morphology.py [hunk 11]
```diff
@@ -49,3 +50,103 @@ def binary_closing(binary_image, kernel):
...
+
+def expand_mask(mask: torch.Tensor, radius: int, threshold: float = 0.5):
+    """Expand a mask by some radius in pytorch"""
+
+    # Needs to be converted to a float to work
+    mask = mask.float()
+
+    # Create a disk-shaped structuring element
+    x, y = torch.meshgrid(
+        torch.arange(-radius, radius + 1), torch.arange(-radius, radius + 1)
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/morphology.py:98]: Future-proof by adding `indexing='ij'`to meshgrid

File: src/home_robot/home_robot/utils/morphology.py [hunk 12]
```diff
@@ -49,3 +50,103 @@ def binary_closing(binary_image, kernel):
...
+        selem.unsqueeze(0).unsqueeze(0),
+        padding=(padding_x, padding_y),
+    )
+
+    # Binarize the expanded mask (optional)
+    expanded_mask = (expanded_mask > 0).to(torch.float32)
+    expanded_mask = expanded_mask[0, 0] > threshold
+    assert (
+        expanded_mask.shape == mask.shape
+    ), "something went wrong when computing padding, most likely - shape not preserved"
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/morphology.py:118]: Nice, informative error message. I like adding the shapes to the output string -- this usually saves me from rerunning the command

File: src/home_robot/home_robot/utils/simple_env.py [hunk 13]
```diff
@@ -5,25 +5,87 @@
...
+        if obs is None:
+            self.reset(seed)
+        else:
+            self.obstacle_pos = obs
+
+    def get_space(self) -> ConfigurationSpace:
+        """Get a space that we can use for planning"""
+        return ConfigurationSpace(2, np.zeros(2), np.ones(2) * self.size, step_size=0.5)
+
+    def validate(self, q):
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/simple_env.py:32]: If this is going to be reused might be helpful to have typehint + shapehint for q

File: src/home_robot/home_robot/utils/simple_env.py [hunk 14]
```diff
@@ -5,25 +5,87 @@
...
     def reset(self, seed: int = 0):
         self.seed = seed
         np.random.seed(seed)
-        # TODO: regenerate start, goal and obstacles
+        self.obstacle_pos = np.random.random(2) * (self.size - self.obstacle_size)
+
+    def show(self, states, backend="mpl", show: bool = True):
+        """Display the scene + states"""
+
+        if not backend == "mlp":
```
Comment (alexsax) [inline src/home_robot/home_robot/utils/simple_env.py:58]: Nice!

File: src/home_robot_hw/home_robot_hw/remote/modules/nav.py [hunk 15]
```diff
@@ -90,18 +96,53 @@ def wait_for_waypoint(
             # Loop until we get there (or time out)
             curr = self.get_base_pose()
             pos_err = np.linalg.norm(xy - curr[:2])
+            rot_err = angle_difference(curr[-1], xyt[2])
             if verbose:
-                print("- curr pose =", curr, "target =", xyt, "err =", pos_err)
-            if pos_err < pos_err_threshold:
+                print(
```
Comment (alexsax) [inline src/home_robot_hw/home_robot_hw/remote/modules/nav.py:101]: If you haven't seen self-documenting f-strings they're pretty convenient! `print(f"- {curr pose=} {target =}...")`

File: tests/home_robot/motion/test_rrt.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (alexsax) [inline tests/home_robot/motion/test_rrt.py:1]: YAY TESTS :)

File: src/home_robot/home_robot/mapping/voxel_map.py [hunk 17]
```diff
@@ -2,25 +2,329 @@
...
+            )
+            if show_all:
+                plt.subplot(8, 8, i + 1)
+                plt.axis("off")
+                plt.imshow(mask.cpu().numpy())
+            self._oriented_masks.append(mask)
+        if show_all:
+            plt.show()
+
+    def distance(self, q0: np.ndarray, q1: np.ndarray) -> float:
```
Comment (alexsax) [inline src/home_robot/home_robot/mapping/voxel_map.py:102]: Seems like this method is used?

**Review by alexsax — APPROVED** [review]
LGTM!

File: src/home_robot/home_robot/mapping/voxel_map.py (hunk 9 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel_map.py:280]: Yeah imo this is super easy to parallelize, just not something i wanted to take on just yet
will probably be a good way to guarantee reachable placement poses later though

File: src/home_robot/home_robot/utils/morphology.py (hunk 10 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/morphology.py:61]: ty

File: src/home_robot/home_robot/utils/morphology.py (hunk 11 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/morphology.py:98]: done

File: src/home_robot/home_robot/mapping/voxel_map.py (hunk 17 above)
Comment (cpaxton) [reply src/home_robot/home_robot/mapping/voxel_map.py:102]: part of the ConfigurationSpace interface

File: src/home_robot/home_robot/utils/simple_env.py (hunk 13 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/simple_env.py:32]: added  but this is just for unit tests basically

File: src/home_robot/home_robot/utils/simple_env.py (hunk 14 above)
Comment (cpaxton) [reply src/home_robot/home_robot/utils/simple_env.py:58]: ty

File: src/home_robot_hw/home_robot_hw/remote/modules/nav.py (hunk 15 above)
Comment (cpaxton) [reply src/home_robot_hw/home_robot_hw/remote/modules/nav.py:101]: had not seen this, super useful
